logger/zap.go

```diff
@@ -0,0 +1,168 @@
+	var logLevel zapcore.Level
+	switch logLevelEnv {
+	case "DEBUG":
+		logLevel = zapcore.DebugLevel
+	case "INFO":
+		logLevel = zapcore.InfoLevel
+	case "WARN":
+		logLevel = zapcore.WarnLevel
+	case "ERROR":
+		logLevel = zapcore.ErrorLevel
+	case "DPANIC":
+		logLevel = zapcore.DPanicLevel
+	case "PANIC":
+		logLevel = zapcore.PanicLevel
+	case "FATAL":
+		logLevel = zapcore.FatalLevel
+	default:
+		logLevel = zapcore.DebugLevel
+	}
+	level := zap.NewAtomicLevel()
+	level.SetLevel(logLevel)
+	cfg := zap.Config{
+		Level:    level,
+		Encoding: "json",
+		EncoderConfig: zapcore.EncoderConfig{
+			TimeKey:        "TimeStamp",
+			LevelKey:       "Severity",
+			NameKey:        "Name",
+			CallerKey:      "Caller",
+			MessageKey:     "Body",
+			EncodeLevel:    zapcore.CapitalLevelEncoder,
+			EncodeTime:     zapcore.ISO8601TimeEncoder,
+			EncodeDuration: zapcore.StringDurationEncoder,
+			EncodeCaller:   zapcore.ShortCallerEncoder,
+		},
+		OutputPaths: []string{"stdout", "/tmp/zap.log"},
+	}
+	logger, err := cfg.Build(
+		zap.AddCallerSkip(2),
+	)
+	if err != nil {
+		log.Fatalf("CreateLogger Error: %v", err)
+	}
+
+	zapLogger = new(ZapLogger)
+	zapLogger.Zap = logger.Sugar()
+
+}
+
+func GetLogger() *ZapLogger {
+	return zapLogger
+}
+
+func (zl *ZapLogger) Errorw(
+	msg string,
+	err error,
+	stackKey string,
+) {
+	zl.Zap.Errorw(
+		msg,
+		zap.Error(err),
+		zap.StackSkip(stackKey, 3),
+	)
+}
+
+func (zl *ZapLogger) ErrorwChannel(
+	msg string,
+	srcChainID, srcChannelID, srcPortID string,
+	dstChainID, dstChannelID, dstPortID string,
+	err error,
+	stackKey string,
+) {
+	zl.Zap.Errorw(
+		msg,
+		"source chain id", srcChainID,
+		"source channnel id", srcChannelID,
+		"source port id", srcPortID,
+		"destination chain id", dstChainID,
+		"destination channel id", dstChannelID,
+		"destination port id", dstPortID,
+		zap.Error(err),
+		zap.StackSkip(stackKey, 3),
+	)
+}
+
+func (zl *ZapLogger) ErrorwConnection(
+	msg string,
+	srcChainID, srcClientID, srcConnectionID,
+	dstChainID, dstClientID, dstConnectionID string,
+	err error,
+	stackKey string,
+) {
+	zl.Zap.Errorw(
+		msg,
+		"source chain id", srcChainID,
+		"source client id", srcClientID,
+		"source connection id", srcConnectionID,
+		"destination chain id", dstChainID,
+		"destination client id", dstClientID,
+		"destination connection id", dstConnectionID,
+		zap.Error(err),
+		zap.StackSkip(stackKey, 3),
+	)
+}
+
+func (zl *ZapLogger) Infow(
+	msg string,
+	info interface{},
+) {
+	zl.Zap.Infow(
+		msg,
+		"info", info,
+	)
+}
+
+func (zl *ZapLogger) InfowChannel(
+	msg string,
+	srcChainID, srcChannelID, srcPortID string,
+	dstChainID, dstChannelID, dstPortID string,
+	info interface{},
+) {
+	zl.Zap.Infow(
+		msg,
+		"source chain id", srcChainID,
+		"source channnel id", srcChannelID,
+		"source port id", srcPortID,
+		"destination chain id", dstChainID,
+		"destination channel id", dstChannelID,
+		"destination port id", dstPortID,
+		"info", info,
+	)
+}
```

Comment:
Why not provide a custom logger instead like below?
```
func GetChannelLogger(
	srcChainID, srcChannelID, srcPortID string,
	dstChainID, dstChannelID, dstPortID string,
) *zap.SugaredLogger {
	return zapLogger.Zap.With(
		"source chain id", srcChainID,
		"source channnel id", srcChannelID,
		"source port id", srcPortID,
		"destination chain id", dstChainID,
		"destination channel id", dstChannelID,
		"destination port id", dstPortID,
        )
}
```

Taking many parameters is a bit annoying, and so it is better to also provide a different version like:
```
func GetChannelLoggerFromChain(chain core.Chain) *zap.SugaredLogger {
	return GetChannelLogger(chain.Path().Src.ChainID, ...)
}
```

Reply:
Please also provide a custom logger like below.
This one it to be used in chain/prover module like `ethereum-ibc-relay-chain`.
```
func GetModuleLogger(
	logger *zap.SugaredLogger,
	moduleName string,
) *zap.SugaredLogger {
	return logger.With("module", moduleName)
}
```

Reply:
Add `GetChannelLogger` `GetChannelLoggerFromChain` `GetModuleLogger `
https://github.com/hyperledger-labs/yui-relayer/pull/92/commits/2cd89343a2eb439b4ab72ae721bf841457c68895

Reply:
@dongrie I'm sorry for failing to convey my intentions.

In the previous review comments, I suggested to create scoped loggers.
My intention was that packing context information (client/connection/channel id, packet sequence, module name, and so on) into scoped loggers would make using Zap's logger functions (`Infow` or `Errorw`) directly painless, in most cases.
In other words, I thought (most of) wrapper functions could be deleted and hoped so.

Of course, a few of the wrapper functions may be still useful.
For example, `ZapLogger::Errorw` is still good since writing `zap.StackSkip(...)` everywhere you logs is annoying.
However, I think other wrapper functions can be deleted by adopting scoped loggers.

Reply:
@siburu Wrapping functions have been removed
https://github.com/hyperledger-labs/yui-relayer/pull/92/commits/d542e085f1819371af8a64d293ab9fe6f229cde7